File: AppSets-iOS/AppSets/com/app/appsets/db/realm/repository/UserInfoRepository.swift
//
//  UserInfoRepository.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/13.
//

import Foundation

class UserInfoRepository {
    
    private static let TAG = "UserInfoRepository"
    
    private let userInfoDao: UserInfoDao = UserInfoDaoImpl()
    
    func addRelatedUserInfoList(_ userInfoList: [UserInfo]) {
        PurpleLogger.current.d(UserInfoRepository.TAG, "addRelatedUserInfoList")
        userInfoDao.addUserInfo(userInfoList)
    }
    
    func getRelatedUserList() -> [UserInfo] {
        PurpleLogger.current.d(UserInfoRepository.TAG, "getRelatedUserList")
        let uids = RelationsUseCase.Instance.getRelatedUserIds()
        return userInfoDao.getUserInfoByUids(uids)
    }
    
    func getUnRelatedUserList() -> [UserInfo] {
        PurpleLogger.current.d(UserInfoRepository.TAG, "getUnRelatedUserList")
        let uids = RelationsUseCase.Instance.getUnRelatedUserIds()
        return userInfoDao.getUserInfoByUids(uids)
    }
    
}

File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/db/room/repository/UserInfoRepository.kt
package xcj.app.appsets.db.room.repository

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import xcj.app.appsets.db.room.dao.UserInfoDao
import xcj.app.appsets.server.model.UserInfo
import xcj.app.appsets.usecase.RelationsUseCase
import xcj.app.appsets.util.PictureUrlMapper
import xcj.app.starter.android.util.PurpleLogger

class UserInfoRepository(
    private val userInfoDao: UserInfoDao
) {

    suspend fun addRelatedUserInfoList(userInfoList: List<UserInfo>) = withContext(Dispatchers.IO) {
        PurpleLogger.current.d(TAG, "addRelatedUserInfoList, thread:${Thread.currentThread()}")
        return@withContext userInfoDao.addUserInfo(*userInfoList.toTypedArray())
    }

    suspend fun getRelatedUserList(): List<UserInfo> = withContext(Dispatchers.IO) {
        PurpleLogger.current.d(TAG, "getRelatedUserList, thread:${Thread.currentThread()}")
        val uids = RelationsUseCase.getInstance().getRelatedUserIds()
        if (uids.isEmpty()) {
            return@withContext emptyList()
        }
        val userInfoList = userInfoDao.getUserInfoByUids(*uids.toTypedArray())
        PictureUrlMapper.mapPictureUrl(userInfoList)
        return@withContext userInfoList
    }

    suspend fun getUnRelatedUserList(): List<UserInfo> = withContext(Dispatchers.IO) {
        PurpleLogger.current.d(TAG, "getUnRelatedUserList, thread:${Thread.currentThread()}")
        val uids = RelationsUseCase.getInstance().getUnRelatedUserIds()
        if (uids.isEmpty()) {
            return@withContext emptyList()
        }
        val userInfoList = userInfoDao.getUserInfoByUids(*uids.toTypedArray())
        PictureUrlMapper.mapPictureUrl(userInfoList)
        return@withContext userInfoList
    }

    companion object {

        private const val TAG = "UserInfoRepository"
        private var INSTANCE: UserInfoRepository? = null

        fun getInstance(): UserInfoRepository {
            if (INSTANCE == null) {
                val repository = UserInfoRepository(UserInfoDao.getInstance())
                INSTANCE = repository
            }
            return INSTANCE!!
        }
    }
}